### PR TITLE
kind/feature : populate-params-and-results-to-workspace-bindings

### PR DESCRIPTION
**Pull request by ericzzzzzzz**
# Changes
fixes: #6384

 - Support for variable interpolation in workspace.* (in PipelineRun and TaskRun) 
 - Note that: AffinityAssistant must be disabled to use tasks.results for pvc in pipeline, as AffinityAssistant is created before all tasks and the pod annotation for affinity only has a place holder, the taskRun creation will fail due to unable to find suitable node. 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Populate params and results to workspace bindings.
```

### CODE REVIEW

**Review by vdemeester — CHANGES_REQUESTED** [review]
Looks good, but I have one question : is it intentional to not support the `csi` workspace binding ? I don't really see any reason to support all (pvc, claimtemplate, secret, configmap and projected) but `csi` 🤔

**Comment by ericzzzzzzz** [comment]
@vdemeester 
> Looks good, but I have one question : is it intentional to not support the `csi` workspace binding ? I don't really see any reason to support all (pvc, claimtemplate, secret, configmap and projected) but `csi` 🤔

ahh, sorry I thought we just wanted the fields listed on the issue. I'll add support for `csi`,

**Review by vdemeester — APPROVED** [review]

**Review by JeromeJu — APPROVED** [review]

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 1]
```diff
@@ -246,6 +246,9 @@ func ApplyTaskResults(targets PipelineRunState, resolvedResultRefs ResolvedResul
 				pipelineTask.TaskRef.Params = pipelineTask.TaskRef.Params.ReplaceVariables(stringReplacements, arrayReplacements, objectReplacements)
 			}
 			pipelineTask.DisplayName = substitution.ApplyReplacements(pipelineTask.DisplayName, stringReplacements)
+			for _, workspace := range pipelineTask.Workspaces {
+				workspace.SubPath = substitution.ApplyReplacements(workspace.SubPath, stringReplacements)
+			}
```
Comment (chitrangpatel) [inline pkg/reconciler/pipelinerun/resources/apply.go:251]: Needs coverage.

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 2]
```diff
@@ -374,6 +377,55 @@ func propagateParams(t v1.PipelineTask, stringReplacements map[string]string, ar
...
+		for _, res := range taskResults {
+			switch res.Type {
+			case v1.ResultsTypeString:
+				stringReplacements[fmt.Sprintf("tasks.%s.results.%s", taskName, res.Name)] = res.Value.StringVal
+			case v1.ResultsTypeArray:
+				continue
+			case v1.ResultsTypeObject:
+				for k, v := range res.Value.ObjectVal {
+					stringReplacements[fmt.Sprintf("tasks.%s.results.%s.%s", taskName, res.Name, k)] = v
+				}
```
Comment (chitrangpatel) [inline pkg/reconciler/pipelinerun/resources/apply.go:394]: Needs coverage

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 3]
```diff
@@ -374,6 +377,55 @@ func propagateParams(t v1.PipelineTask, stringReplacements map[string]string, ar
...
+		}
+		pr.Spec.Workspaces[i].SubPath = substitution.ApplyReplacements(binding.SubPath, stringReplacements)
+		if pr.Spec.Workspaces[i].ConfigMap != nil {
+			pr.Spec.Workspaces[i].ConfigMap.Name = substitution.ApplyReplacements(binding.ConfigMap.Name, stringReplacements)
+		}
+		if pr.Spec.Workspaces[i].CSI != nil {
+			pr.Spec.Workspaces[i].CSI.Driver = substitution.ApplyReplacements(binding.CSI.Driver, stringReplacements)
+			if pr.Spec.Workspaces[i].CSI.NodePublishSecretRef != nil {
+				pr.Spec.Workspaces[i].CSI.NodePublishSecretRef.Name = substitution.ApplyReplacements(binding.CSI.NodePublishSecretRef.Name, stringReplacements)
+			}
```
Comment (chitrangpatel) [inline pkg/reconciler/pipelinerun/resources/apply.go:411]: needs coverage.

**Review by chitrangpatel — CHANGES_REQUESTED** [review]
Almost lgtm. Just need coverage on the places I highlighted I think. They are untested.

**Pull request merged**